File: gnu/gcc/libgomp/testsuite/libgomp.fortran/appendix-a/a.2.1.f90
! { dg-do run }
PROGRAM A2
  INCLUDE "omp_lib.h"      ! or USE OMP_LIB
  INTEGER X
  X=2
!$OMP PARALLEL NUM_THREADS(2) SHARED(X)
    IF (OMP_GET_THREAD_NUM() .EQ. 0) THEN
       X=5
    ELSE
    ! PRINT 1: The following read of x has a race
      PRINT *,"1: THREAD# ", OMP_GET_THREAD_NUM(), "X = ", X
    ENDIF
!$OMP BARRIER
    IF (OMP_GET_THREAD_NUM() .EQ. 0) THEN
    ! PRINT 2
      PRINT *,"2: THREAD# ", OMP_GET_THREAD_NUM(), "X = ", X
    ELSE
    ! PRINT 3
      PRINT *,"3: THREAD# ", OMP_GET_THREAD_NUM(), "X = ", X
    ENDIF
!$OMP END PARALLEL
END PROGRAM A2

File: gnu/gcc/libgomp/testsuite/libgomp.c/appendix-a/a.2.1.c
/* { dg-do run } */

#include <stdio.h>
#include <omp.h>
extern void abort (void);
int
main ()
{
  int bad, x;
  x = 2;
  bad = 0;
#pragma omp parallel num_threads(2) shared(x, bad)
  {
    if (omp_get_thread_num () == 0)
      {
	volatile int i;
	for (i = 0; i < 100000000; i++)
	  x = 5;
      }
    else
      {
	/* Print 1: the following read of x has a race */
	if (x != 2 && x != 5)
	  bad = 1;
      }
#pragma omp barrier
    if (omp_get_thread_num () == 0)
      {
	/* x must be 5 now.  */
	if (x != 5)
	  bad = 1;
      }
    else
      {
	/* x must be 5 now.  */
	if (x != 5)
	  bad = 1;
      }
  }

  if (bad)
    abort ();

  return 0;
}
